### PR TITLE
Specify URL equivalence modulo search variance

### LINKED ISSUE #201: No-Vary-Search canonicalization behavior
No-Vary-Search does not have a processing model for actually matching cache keys specified yet.

However, our plan is to base the processing model on [the application/x-www-form-urlencoded parser](https://url.spec.whatwg.org/#urlencoded-parsing).

This leads to some interesting behavior, because even if two strings are exact matches in terms of parsed search params, they may not be exact string matches. For example, `/foo?bar=baz&&&&` and `/foo?bar=baz` both have the same parsed search params.

The question is, when do these start matching?

- Does `No-Vary-Search: params=?0, key-order=?0`, which looks like it should have no effect, actually cause that new non-default behavior to occur?

- Do we decide to start doing this only after `params` becomes nonempty, or `key-order` gets set to true?

- Other ideas?

### PR DESCRIPTION
**Pull request by domenic**
Closes #201.

@liviutinta FYI

### CODE REVIEW

File: no-vary-search.bs [hunk 1]
```diff
@@ -168,3 +168,57 @@ The [=obtain a URL search variance=] algorithm ensures that all [=URL search var
...
+
+1. If |searchVariance|'s [=URL search variance/vary on key order=] is false, then:
+
+  1. Let |keyLessThan| be an algorithm taking as inputs two pairs (|keyA|, <var ignore>valueA</var>) and (|keyB|, <var ignore>valueB</var>), which returns whether |keyA| is [=code unit less than=] |keyB|.
+
+  1. Set |searchParamsA| to the result of [=list/sorting in ascending order=] |searchParamsA|, with |keyLessThan|.
+
+  1. Set |searchParamsB| to the result of [=list/sorting in ascending order=] |searchParamsB|, with |keyLessThan|.
+
+1. If |searchParamsA|'s [=list/size=] is not equal to |searchParamsB|'s [=list/size=], then return false.
```
Comment (domenic) [inline no-vary-search.bs:212]: I was contemplating glossing over the following lines with a hand-waved "deep equal", but I dunno...

Comment (jeremyroman) [reply no-vary-search.bs:212]: This works for me.

**Review by jeremyroman — APPROVED** [review]

File: no-vary-search.bs [hunk 2]
```diff
@@ -168,3 +168,57 @@ The [=obtain a URL search variance=] algorithm ensures that all [=URL search var
...
+
+  1. If |urlA|'s [=url/query=] equals |urlB|'s [=url/query=], then return true.
+
+  1. Return false.
+
+  <p class="note">In this case, even [=URL=] pairs that might appear the same after parsing, such as `https://example.com/a` and `https://example.com/a?`, or `https://example.com/foo?a=b&&&c` and `https://example.com/foo?a=b&c=`, will be treated as inequivalent.
+
+1. Let |searchParamsA| and |searchParamsB| be empty [=lists=].
+
+1. If |urlA|'s [=url/query=] is not null, then set |searchParamsA| to the result of running the [=urlencoded parser|application/x-www-form-urlencoded parser=] given the [=isomorphic encoding=] of |urlA|'s [=url/query=].
```
[the remark this answers is not in the document]
Comment (domenic) [reply no-vary-search.bs:208]: I added a table of such examples. Let me know if I got any wrong, or if there are others I should include.

[the remark this answers is not in the document]
Comment (jeremyroman) [reply no-vary-search.bs:208]: Yeah, I think that's great. We could technically also call out the implication for keys as well as values, but perhaps that's overkill and can be inferred from this.

**Pull request merged**